src/main/java/io/opentracing/contrib/global/GlobalTracer.java

```diff
@@ -0,0 +1,148 @@
+ * <li>Automatically by using the Java <code>ServiceLoader</code> SPI mechanism to load an implementation
+ * from the classpath.</li>
+ * </ol>
+ */
+public final class GlobalTracer implements Tracer {
+    private static final Logger LOGGER = Logger.getLogger(GlobalTracer.class.getName());
+
+    private static final GlobalTracer INSTANCE = new GlobalTracer();
+
+    /**
+     * The resolved {@link Tracer} to delegate the global tracing implementation to.<br>
+     * This can be either an {@link #set(Tracer) explicitly set delegate}
+     * or the automatically resolved Tracer implementation.
+     */
+    private final AtomicReference<Tracer> globalTracer = new AtomicReference<Tracer>();
+
+    private GlobalTracer() {
+    }
+
+    private Tracer lazyTracer() {
+        Tracer instance = globalTracer.get();
+        if (instance == null) {
+            final Tracer singleton = loadSingleton();
+            while (instance == null && singleton != null) { // for race condition only, should rarely happen.
+                globalTracer.compareAndSet(null, singleton);
+                instance = globalTracer.get();
+            }
+            LOGGER.log(Level.INFO, "Using GlobalTracer implementation: {0}.", instance);
+        }
+        LOGGER.log(Level.FINEST, "GlobalTracer: {0}.", instance);
+        return instance;
+    }
+
+    /**
+     * Returns the {@linkplain GlobalTracer} instance.
+     * Upon first use of any tracing method, this tracer lazily determines which actual {@link Tracer}
+     * implementation to use:
+     * <ol type="a">
+     * <li>If an explicitly configured tracer was provided via the {@link #set(Tracer)} method,
+     * that will always take precedence over automatically provided tracer instances.</li>
+     * <li>A Tracer implementation can be automatically provided using the Java {@link ServiceLoader} through the
+     * <code>META-INF/services/io.opentracing.Tracer</code> service definition file.<br>
+     * The {@linkplain GlobalTracer} will not attempt to choose between implementations;
+     * if more than one is found by the {@linkplain ServiceLoader service loader},
+     * a warning is logged and tracing is disabled by falling back to the default implementation:</li>
+     * <li>If no tracer implementation is available, the {@link io.opentracing.NoopTracer NoopTracer}
+     * will be used.</li>
+     * </ol>
+     *
+     * @return The global tracer.
+     */
+    public static Tracer get() {
+        return INSTANCE;
+    }
+
+    /**
+     * Explicit registration of a configured {@link Tracer} to back the behaviour
+     * of the {@link #get() global tracer}.
+     * <p>
+     * The previous global tracer is returned so it can be restored later if necessary.
+     *
+     * @param tracer Tracer to use as global tracer.
+     * @return The previous global tracer.
+     */
+    public static Tracer set(final Tracer tracer) {
```

Comment:
@sjoerdtalsma I am ok merging this (i.e., I don't need to win every argument), but one approach to the asymmetry problem would be to keep the semantics but simply change the name – IMO that would be a lot clearer for a new user / reader. Something like `GlobalTracer.instance()` instead of `GlobalTracer.get()`, for example. I could also imagine `GlobalTracer.install(tracer)` instead of `GlobalTracer.set(tracer)`.

Let me know what you think. If there are no other objections I'm ok merging this soon, though.